language/libra-vm/src/errors.rs

```diff
@@ -0,0 +1,147 @@
+pub fn convert_normal_prologue_error(status: VMStatus) -> VMStatus {
+    match status {
+        VMStatus::Executed => VMStatus::Executed,
+        VMStatus::MoveAbort(location, code) => {
+            if location != known_locations::account_module_abort() {
+                crit!(
+                    "[libra_vm] Unexpected prologue move abort: {:?}::{:?}",
+                    location,
+                    code
+                );
+                return VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION);
+            }
+
+            let new_major_status = match code {
+                EACCOUNT_FROZEN => StatusCode::SENDING_ACCOUNT_FROZEN,
+                // Invalid authentication key
+                EBAD_ACCOUNT_AUTHENTICATION_KEY => StatusCode::INVALID_AUTH_KEY,
+                // Sequence number too old
+                ESEQUENCE_NUMBER_TOO_OLD => StatusCode::SEQUENCE_NUMBER_TOO_OLD,
+                // Sequence number too new
+                ESEQUENCE_NUMBER_TOO_NEW => StatusCode::SEQUENCE_NUMBER_TOO_NEW,
+                // Sequence number too new
+                EACCOUNT_DOES_NOT_EXIST => StatusCode::SENDING_ACCOUNT_DOES_NOT_EXIST,
+                // Can't pay for transaction gas deposit/fee
+                EINSUFFICIENT_BALANCE => StatusCode::INSUFFICIENT_BALANCE_FOR_TRANSACTION_FEE,
+                ETRANSACTION_EXPIRED => StatusCode::TRANSACTION_EXPIRED,
+                EBAD_CHAIN_ID => StatusCode::BAD_CHAIN_ID,
+                code => {
+                    crit!(
+                        "[libra_vm] Unexpected prologue move abort: {:?}::{:?}",
+                        location,
+                        code
+                    );
+                    return VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION);
+                }
+            };
+            VMStatus::Error(new_major_status)
+        }
+        status @ VMStatus::ExecutionFailure { .. } | status @ VMStatus::Error(_) => {
+            crit!("[libra_vm] Unexpected prologue error: {:?}", status);
+            VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION)
+        }
+    }
+}
+
+/// Checks for only move aborts or successful execution.
+/// Any other errors are mapped to the invariant violation
+/// `UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION`
+pub fn convert_normal_success_epilogue_error(status: VMStatus) -> VMStatus {
+    match status {
+        VMStatus::MoveAbort(location, code @ EACCOUNT_FROZEN)
+        | VMStatus::MoveAbort(location, code @ EINSUFFICIENT_BALANCE) => {
+            if location != known_locations::account_module_abort() {
+                crit!(
+                    "[libra_vm] Unexpected success epilogue move abort: {:?}::{:?}",
+                    location,
+                    code
+                );
+                return VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION);
+            }
+            VMStatus::MoveAbort(location, code)
```

Comment:
Should these 2 MoveAbort codes be translated to VMStatus::Error with the corresponding StatusCode::SENDING_ACCOUNT_FROZEN and StatusCode::INSUFFICIENT_BALANCE_FOR_TRANSACTION_FEE codes? This is the only case where prologue/epilogue aborts are returned as VMStatus::MoveAbort, so if this is really intentional, it would be good to have a comment here explaining why this is different than the other places.

Reply:
Not really sure. Would be a good question for maybe @sblackshear @runtian-zhou or @tzakian 
My guess is that the other ones are converted so its easy on the validator (we could get rid of the conversion if we really wanted). If the success epilogue aborts, we don't really care why, we just want to then call the failure epilogue